### ISSUE TITLE
[BUG] needs dependency graph is broken

### ISSUE THREAD

**Issue by dudicoco** [roboll/helmfile#1881]
With helmfile v0.139.9 it seems that the `needs` dependencies behaviour is broken.

With the following example:
```yaml
- name: amazing-app-test-1
  chart: my-repo/amazing-app
  version: 2.0.14
  installed: true
  timeout: 300
  needs:
  - devops/amazing-app-test-3
  values: []
- name: amazing-app-test-2
  chart: my-repo/amazing-app
  version: 2.0.14
  installed: true
  timeout: 300
  needs:
  - devops/amazing-app-test-1
  values: []
- name: amazing-app-test-3
  chart: my-repo/amazing-app
  version: 2.0.14
  installed: true
  timeout: 300
  needs: []
  values: []
```

Previous behaviour - v0.138.4 (correct):
```
Upgrading release=amazing-app-test-3, chart=my-repo/amazing-app
Release "amazing-app-test-3" does not exist. Installing it now.
NAME: amazing-app-test-3
LAST DEPLOYED: Mon Jun 14 10:55:29 2021
NAMESPACE: devops
STATUS: deployed
REVISION: 1
TEST SUITE: None

Listing releases matching ^amazing-app-test-3$
amazing-app-test-3	devops   	1       	2021-06-14 10:55:29.884771 +0300 IDT	deployed	amazing-app-2.0.14	1.0

Upgrading release=amazing-app-test-1, chart=my-repo/amazing-app
Release "amazing-app-test-1" does not exist. Installing it now.
NAME: amazing-app-test-1
LAST DEPLOYED: Mon Jun 14 10:56:28 2021
NAMESPACE: devops
STATUS: deployed
REVISION: 1
TEST SUITE: None

Listing releases matching ^amazing-app-test-1$
amazing-app-test-1	devops   	1       	2021-06-14 10:56:28.275669 +0300 IDT	deployed	amazing-app-2.0.14	1.0

Upgrading release=amazing-app-test-2, chart=my-repo/amazing-app
Release "amazing-app-test-2" does not exist. Installing it now.
NAME: amazing-app-test-2
LAST DEPLOYED: Mon Jun 14 10:57:09 2021
NAMESPACE: devops
STATUS: deployed
REVISION: 1
TEST SUITE: None

Listing releases matching ^amazing-app-test-2$
amazing-app-test-2	devops   	1       	2021-06-14 10:57:09.207185 +0300 IDT	deployed	amazing-app-2.0.14	1.0
```

New behaviour - v0.139.9 (incorrect - releases 1 & 2 are installed in parallel):
```
Upgrading release=amazing-app-test-3, chart=my-repo/amazing-app
Release "amazing-app-test-3" does not exist. Installing it now.
NAME: amazing-app-test-3
LAST DEPLOYED: Mon Jun 14 11:01:38 2021
NAMESPACE: devops
STATUS: deployed
REVISION: 1
TEST SUITE: None

Listing releases matching ^amazing-app-test-3$
amazing-app-test-3	devops   	1       	2021-06-14 11:01:38.096974 +0300 IDT	deployed	amazing-app-2.0.14	1.0

Upgrading release=amazing-app-test-2, chart=my-repo/amazing-app
Upgrading release=amazing-app-test-1, chart=my-repo/amazing-app
Release "amazing-app-test-1" does not exist. Installing it now.
NAME: amazing-app-test-1
LAST DEPLOYED: Mon Jun 14 11:02:16 2021
NAMESPACE: devops
STATUS: deployed
REVISION: 1
TEST SUITE: None

Listing releases matching ^amazing-app-test-1$
Release "amazing-app-test-2" does not exist. Installing it now.
NAME: amazing-app-test-2
LAST DEPLOYED: Mon Jun 14 11:02:16 2021
NAMESPACE: devops
STATUS: deployed
REVISION: 1
TEST SUITE: None

Listing releases matching ^amazing-app-test-2$
amazing-app-test-1	devops   	1       	2021-06-14 11:02:16.553073 +0300 IDT	deployed	amazing-app-2.0.14	1.0

amazing-app-test-2	devops   	1       	2021-06-14 11:02:16.419325 +0300 IDT	deployed	amazing-app-2.0.14	1.0
```

**Comment by paologallinaharbur** [comment]
I noticed the very same behaviour, it seemed that the `need` option was completely ignored 😕

**Comment by kbbqiu** [comment]
Running into this as well.

**Comment by dudicoco** [comment]
@mumoshu can you please address this?

**Comment by mumoshu** [comment]
Sorry everyone I've been occupied by other matters these days so I won't be able to work on this shortly.

It would be great if anyone could stand up and submit a PR for this. I'll gladly review any PR related to this ASAP.

**Comment by dm3ch** [comment]
Maybe it's caused by this https://github.com/roboll/helmfile/releases/tag/v0.139.2 ?

**Comment by rafilkmp3** [comment]
can be enabled as a default with

```
helmDefaults:
  skipDeps: false
```

**Comment by dudicoco** [comment]
`--include-needs` or `--skip-needs=false` will not solve this issue.

@mumoshu the whole `include-needs` feature get reverted in order to fix this bug?
We are currently blocked from upgrading our helmfile version.

**Comment by mumoshu** [comment]
@dm3ch Ah that's a good point! Maybe we should enable `--include-needs` by default, rather than enabling `--skip-needs`?

If anyone could submit a feature to change the default, like I've done for `--skip-needs` in https://github.com/roboll/helmfile/pull/1835, I would gladly review it.

**Comment by mumoshu** [comment]
@dudicoco Does it have something to do with you're missing `namespace: devops` in your releases, even though you refer to dependencies with namespaces like `devops/amazing-app-test-3`? My understanding was that you can use namespacced dependency only when you have namespace defined in the dependeded release.

**Comment by dudicoco** [comment]
@mumoshu we have one helmfile.yaml file which loads all of the releases and templates them, during the templating the namespace is added, so this is not the issue here.
This has worked for us in `v0.138.4`.

`--include-needs` will not solve the issue, as we do not want to force add releases defined in the needs block unless they were explicitly included via `--label`.

**Comment by mumoshu** [comment]
@dm3ch Well would you mind clarifying what are you expecting from `--include-needs` vs `--skip-needs`, exactly? They controls how Helmfile handles indirectly selected releases, to be installed along with directly selected releases, or to be skipped.

**Comment by mumoshu** [comment]
@dudicoco Gotcha! Thanks. Do you have any blocker that prevents you from fixing Helmfile in the main branch(as I currently lack spare time to work on Helmfile) OR sticking with `v0.138.4`?

**Comment by dudicoco** [comment]
@mumoshu can you please elaborate on what you mean by fixing Helmfile in the main branch?

On a side note, perhaps additional maintainers should be added to the repository? There is a high traffic on this repository and I understand the difficulty in maintaining it all by yourself :)

Thanks

**Comment by mumoshu** [comment]
> can you please elaborate on what you mean by fixing Helmfile in the main branch?

@dudicoco I mean "what is preventing you from submitting a pull request yourself" 😃 

> On a side note, perhaps additional maintainers should be added to the repository? There is a high traffic on this repository and I understand the difficulty in maintaining it all by yourself :)

Yes, I and @renehernandez have already done our best at #1824 and a few Slack channels but had no luck so far. I don't have admin privilege on this repo so I can't add additional maintainers as of today.

**Comment by mumoshu** [comment]
FYI, we have already a lot of unit tests like https://github.com/roboll/helmfile/blob/master/pkg/app/app_test.go#L2749 that are for avoiding regressions like this. At least unit tests are passing. Especially the `smoke` test covers multiple releases with a complex DAG so `needs` should work. Apparently there're still edge-cases?

**Comment by dudicoco** [comment]
@mumoshu if it's a PR for fixing the bug, I don't think i'll be able to find and fix the problem within the code :)
If it's for reverting the `include-needs` feature, I could give it a shot.

Regarding the smoke test, the use case here is very simple and straightforward use of `needs` so I don't believe it's an edge-case, I've noticed that the smoke test uses releases without namespace prefixed, perhaps this is why they are passing?

**Comment by mumoshu** [comment]
I recently verified `needs` to work while verifying another issue #1959. For #1959, it was that the need is silently ignored if you had any typo in namespace or name of the needed release. So I added an explicit error in that case, in #2026.

Everyone, would you mind sharing a single helmfile.yaml that `needs` doesn't work with?
It's already a half year since this issue has been created but I didn't receive any single helmfile.yaml for reproduction.

The first example provided by @dudicoco in this issue lacked `namespace` in the needed release but he says it's a kind of a base template that is source of the final helmfile.yaml 🤔 https://github.com/roboll/helmfile/issues/1881#issuecomment-886570194

**Comment by dudicoco** [comment]
Hi @mumoshu, here's a more complete `helmfile.yaml` for reproduction:

```yaml
- name: amazing-app-test-1
  chart: my-repo/amazing-app
  kubeContext: my-context
  namespace: devops
  version: 2.0.14
  installed: true
  timeout: 300
  needs:
  - my-context/devops/amazing-app-test-3
  values: []
- name: amazing-app-test-2
  chart: my-repo/amazing-app
  kubeContext: my-context
  namespace: devops
  version: 2.0.14
  installed: true
  timeout: 300
  needs:
  - my-context/devops/amazing-app-test-1
  values: []
- name: amazing-app-test-3
  chart: my-repo/amazing-app
  kubeContext: my-context
  namespace: devops
  version: 2.0.14
  installed: true
  timeout: 300
  needs: []
  values: []
```

Let me know if anything else is needed.

**Comment by mumoshu** [comment]
@dudicoco Thanks! FWIW, I couldn't reproduce the problem with helmfile 0.139.9

```
Upgrading release=amazing-app-test-3, chart=my-repo/amazing-app
Release "amazing-app-test-3" does not exist. Installing it now.
NAME: amazing-app-test-3
LAST DEPLOYED: Mon Jan 10 06:07:46 2022
NAMESPACE: devops
STATUS: deployed
REVISION: 1
NOTES:
1. Get the application URL by running these commands:
  export POD_NAME=$(kubectl get pods --namespace devops -l "app.kubernetes.io/name=amazing-app,app.kubernetes.io/instance=amazing-app-test-3" -o jsonpath="{.items[0].metadata.name}")
  export CONTAINER_PORT=$(kubectl get pod --namespace devops $POD_NAME -o jsonpath="{.spec.containers[0].ports[0].containerPort}")
  echo "Visit http://127.0.0.1:8080 to use your application"
  kubectl --namespace devops port-forward $POD_NAME 8080:$CONTAINER_PORT

Listing releases matching ^amazing-app-test-3$
amazing-app-test-3      devops          1               2022-01-10 06:07:46.038450005 +0000 UTC  deployed        amazing-app-0.1.01.16.0     

Upgrading release=amazing-app-test-1, chart=my-repo/amazing-app
Release "amazing-app-test-1" does not exist. Installing it now.
NAME: amazing-app-test-1
LAST DEPLOYED: Mon Jan 10 06:07:46 2022
NAMESPACE: devops
STATUS: deployed
REVISION: 1
NOTES:
1. Get the application URL by running these commands:
  export POD_NAME=$(kubectl get pods --namespace devops -l "app.kubernetes.io/name=amazing-app,app.kubernetes.io/instance=amazing-app-test-1" -o jsonpath="{.items[0].metadata.name}")
  export CONTAINER_PORT=$(kubectl get pod --namespace devops $POD_NAME -o jsonpath="{.spec.containers[0].ports[0].containerPort}")
  echo "Visit http://127.0.0.1:8080 to use your application"
  kubectl --namespace devops port-forward $POD_NAME 8080:$CONTAINER_PORT

Listing releases matching ^amazing-app-test-1$
amazing-app-test-1      devops          1               2022-01-10 06:07:46.682626195 +0000 UTC  deployed        amazing-app-0.1.01.16.0     

Upgrading release=amazing-app-test-2, chart=my-repo/amazing-app
Release "amazing-app-test-2" does not exist. Installing it now.
NAME: amazing-app-test-2
LAST DEPLOYED: Mon Jan 10 06:07:47 2022
NAMESPACE: devops
STATUS: deployed
REVISION: 1
NOTES:
1. Get the application URL by running these commands:
  export POD_NAME=$(kubectl get pods --namespace devops -l "app.kubernetes.io/name=amazing-app,app.kubernetes.io/instance=amazing-app-test-2" -o jsonpath="{.items[0].metadata.name}")
  export CONTAINER_PORT=$(kubectl get pod --namespace devops $POD_NAME -o jsonpath="{.spec.containers[0].ports[0].containerPort}")
  echo "Visit http://127.0.0.1:8080 to use your application"
  kubectl --namespace devops port-forward $POD_NAME 8080:$CONTAINER_PORT

Listing releases matching ^amazing-app-test-2$
amazing-app-test-2      devops          1               2022-01-10 06:07:47.335290788 +0000 UTC  deployed        amazing-app-0.1.01.16.0     


UPDATED RELEASES:
NAME                 CHART                 VERSION
amazing-app-test-3   my-repo/amazing-app     0.1.0
amazing-app-test-1   my-repo/amazing-app     0.1.0
amazing-app-test-2   my-repo/amazing-app     0.1.0
```

**Comment by dudicoco** [comment]
@mumoshu what command did you use to install the releases?

I've used `helmfile -e my-environment -l name=amazing-app-test-1 -l name=amazing-app-test-2 -l name=amazing-app-test-3 apply`

**Comment by mumoshu** [comment]
@dudicoco I just ran plain `helmfile apply`. I also ran it with `-l` like the below but it seems to work:

```
./helmfile-0.139.9 -f helmfile.1881.yaml -l name=amazing-app-test-1 -l name=amazing-app-test-2 -l name=amazing-app-test-3 apply
```

<details>

```
Upgrading release=amazing-app-test-3, chart=my-repo/amazing-app
Release "amazing-app-test-3" does not exist. Installing it now.
NAME: amazing-app-test-3
LAST DEPLOYED: Mon Jan 10 12:37:36 2022
NAMESPACE: devops
STATUS: deployed
REVISION: 1
NOTES:
1. Get the application URL by running these commands:
  export POD_NAME=$(kubectl get pods --namespace devops -l "app.kubernetes.io/name=amazing-app,app.kubernetes.io/instance=amazing-app-test-3" -o jsonpath="{.items[0].metadata.name}")
  export CONTAINER_PORT=$(kubectl get pod --namespace devops $POD_NAME -o jsonpath="{.spec.containers[0].ports[0].containerPort}")
  echo "Visit http://127.0.0.1:8080 to use your application"
  kubectl --namespace devops port-forward $POD_NAME 8080:$CONTAINER_PORT

Listing releases matching ^amazing-app-test-3$
amazing-app-test-3      devops          1               2022-01-10 12:37:36.013775617 +0000 UTC  deployed        amazing-app-0.1.01.16.0     

Upgrading release=amazing-app-test-1, chart=my-repo/amazing-app
Release "amazing-app-test-1" does not exist. Installing it now.
NAME: amazing-app-test-1
LAST DEPLOYED: Mon Jan 10 12:37:36 2022
NAMESPACE: devops
STATUS: deployed
REVISION: 1
NOTES:
1. Get the application URL by running these commands:
  export POD_NAME=$(kubectl get pods --namespace devops -l "app.kubernetes.io/name=amazing-app,app.kubernetes.io/instance=amazing-app-test-1" -o jsonpath="{.items[0].metadata.name}")
  export CONTAINER_PORT=$(kubectl get pod --namespace devops $POD_NAME -o jsonpath="{.spec.containers[0].ports[0].containerPort}")
  echo "Visit http://127.0.0.1:8080 to use your application"
  kubectl --namespace devops port-forward $POD_NAME 8080:$CONTAINER_PORT

Listing releases matching ^amazing-app-test-1$
amazing-app-test-1      devops          1               2022-01-10 12:37:36.658368508 +0000 UTC  deployed        amazing-app-0.1.01.16.0     

Upgrading release=amazing-app-test-2, chart=my-repo/amazing-app
Release "amazing-app-test-2" does not exist. Installing it now.
NAME: amazing-app-test-2
LAST DEPLOYED: Mon Jan 10 12:37:37 2022
NAMESPACE: devops
STATUS: deployed
REVISION: 1
NOTES:
1. Get the application URL by running these commands:
  export POD_NAME=$(kubectl get pods --namespace devops -l "app.kubernetes.io/name=amazing-app,app.kubernetes.io/instance=amazing-app-test-2" -o jsonpath="{.items[0].metadata.name}")
  export CONTAINER_PORT=$(kubectl get pod --namespace devops $POD_NAME -o jsonpath="{.spec.containers[0].ports[0].containerPort}")
  echo "Visit http://127.0.0.1:8080 to use your application"
  kubectl --namespace devops port-forward $POD_NAME 8080:$CONTAINER_PORT

Listing releases matching ^amazing-app-test-2$
amazing-app-test-2      devops          1               2022-01-10 12:37:37.278351885 +0000 UTC  deployed        amazing-app-0.1.01.16.0     


UPDATED RELEASES:
NAME                 CHART                 VERSION
amazing-app-test-3   my-repo/amazing-app     0.1.0
amazing-app-test-1   my-repo/amazing-app     0.1.0
amazing-app-test-2   my-repo/amazing-app     0.1.0
```
</details>

**Comment by dudicoco** [comment]
@mumoshu i've just tried running it many times with different configurations in my helmfile.yaml and the issue did not go away.

Could there be a difference between our helmfile versions? Is there any other information I can provide you with? perhaps debug logs?

**Comment by mumoshu** [comment]
@dudicoco Would you mind sharing me a github repo with a complete example of a helmfile project that can reproduce the issue?

I don't actually have access to `my-repo/amazing-app` in your example so my example is like:

```
releases:
- name: amazing-app-test-1
  chart: ./my-repo/amazing-app
  kubeContext: my-context
  namespace: devops
  version: 2.0.14
  installed: true
  timeout: 300
  needs:
  - my-context/devops/amazing-app-test-3
  values: []
- name: amazing-app-test-2
  chart: ./my-repo/amazing-app
  kubeContext: my-context
  namespace: devops
  version: 2.0.14
  installed: true
  timeout: 300
  needs:
  - my-context/devops/amazing-app-test-1
  values: []
- name: amazing-app-test-3
  chart: ./my-repo/amazing-app
  kubeContext: my-context
  namespace: devops
  version: 2.0.14
  installed: true
  timeout: 300
  needs: []
  values: []

```

**Comment by dudicoco** [comment]
@mumoshu please try it with chart `bitnami/nginx` version `9.7.1`: https://artifacthub.io/packages/helm/bitnami/nginx

I'm using helmfile version v0.139.9 on darwin x86_64.

**Comment by dudicoco** [comment]
Hi @mumoshu, how can we track this issue in the new repo? Should I just reopen it there?
